docs/development/m4_fpga_communication.md

```diff
@@ -0,0 +1,97 @@
+
+### STATUS
+
+![Layout of STATUS register](/images/reg-status.png)
+
+This register is expected to be informational only, it doesn't provide any
+information that isn't conveyed through other means (interrupts).
+
+* `exec` indicates that MCU is expected to act upon request sent from FPGA,
+  specified in `OP_TYPE` register. [Interrupt](#interrupts) is generated when
+  this signal gets set by FPGA.
+* `abort` is set when PC aborts command currently being executed by TPM.
+  [Interrupt](#interrupts) is generated when this signal gets set by FPGA.
+* `complete` is a signal sent towards FPGA by writing to `COMPLETE` register.
+  The signal is active for a few cycles after write to that register, after that
+  is should be automatically set back to 0.
+
+### OP_TYPE
+
+![Layout of OP_TYPE register](/images/reg-op-type.png)
+
+* `op_type` is a type of operation expected from MCU. 0 is used as a default
+  value to which this register returns after `complete` signal is acknowledged
+  by FPGA. 0xC is reserved as it may be part of 0xBADFABAC magic value. This
+  register is only valid if `exec` is set.
+
+| op_type | Operation                                  |
+|--------:|--------------------------------------------|
+| 0x0     | No operation                               |
+| 0x1     | Execute TPM2 command located in `FPGA_RAM` |
+| 0xC     | Illegal operation type                     |
+| other   | Reserved for future use                    |
+
+### LOCALITY
+
+![Layout of LOCALITY register](/images/reg-locality.png)
+
+* `locality` specifies the locality at which the current operation was
+  initiated. Value 0xF is used when no locality was active at that point. This
+  field is only valid if `exec` is set.
+
+### BUF_SIZE
+
+![Layout of BUF_SIZE register](/images/reg-buf-size.png)
+
+* `buf_size` specifies size (in bytes) of data available in FPGA RAM, which is
+  usually a TPM2 command. This field is only valid if `exec` is set. TPM stack
+  is not expected to write size of response to this field, it is deduced from
+  response's payload.
+
+### COMPLETE
+
+Any write to this register triggers `complete` signal to be sent to FPGA. Reads
+from this register aren't implemented, meaning that 0xBADFABAC is returned.
+
+## FPGA RAM
+
+This region behaves like RAM if `exec` is set. Its content depends on `op_type`,
+which in most cases will be TPM2 command. Response is written by MCU to this
+buffer, starting from offset 0, overwriting the command.
+
+When `exec` is not set, data read from this region is invalid and writes are
+dropped.
+
+## Interrupts
+
+TwPM uses two interrupt signals from FPGA towards M4. First one is `exec` which
+signals that there is an operation to be performed by TPM stack, and another one
+tells that the currently running operation is to be aborted.
+
+Both interrupts arrive to M4 as IRQ 4, and exact reason for interrupt can be
+read from 4-byte `FB_INTR_RAW` NVIC register located at address 0x40004884. Bits
```

Comment:
@krystian-hebel This is not part of NVIC. This is part of something called `Intr_Ctrl`, not documented in EOS S3 TRM. But documented in https://github.com/QuickLogic-Corp/EOS-S3/blob/master/Docs/EOSS3-Registers.ods.

Reply:
Fixed in https://github.com/Dasharo/twpm-docs/pull/15/commits/86d5b7dfaf5b7af1f6ce41a72798a338cfb3b72c